taskfiles/utils/cpp-lint.yaml

```diff
@@ -4,36 +4,324 @@ set: ["u", "pipefail"]
+  # On failure the `clang-tidy` output is printed. The task will fail on the first file to fail
+  # `clang-tidy`, which will result in some files having incomplete output. Either fix the known
+  # errors and run the task again or directly run `clang-tidy` on a file of interest.
+  #
+  # @param {string} VENV_DIR Python virtual environment where `clang-tidy` is installed.
+  # @param {string} [GIT_REF="origin/main"] A `git` reference that HEAD is diff'd with.
+  # @param {string[]} [FLAGS] Any flags to pass to `clang-tidy`.
+  # @param {string} [OUTPUT_DIR="{{.ROOT_DIR}}/lint-clang-tidy"] Directory to store the `clang-tidy`
+  # output for each file.
+  clang-tidy-diff:
+    internal: true
+    label: "{{.TASK}}:{{.OUTPUT_DIR}}"
+    vars:
+      GIT_REF: >-
+        {{default "origin/main" .GIT_REF}}
+      DIFF_SOURCE_PATHS_:
+        sh: >-
+          git diff --name-only --ignore-submodules "{{.GIT_REF}}"
+          | grep ".*\.\(cpp\|h\|hpp\)"
+          | xargs --no-run-if-empty realpath
+      DIFF_SOURCE_PATHS:
+        ref: >-
+          splitList "\n" .DIFF_SOURCE_PATHS_
+      FLAGS:
+        ref: "default (list) .FLAGS"
+      OUTPUT_DIR: >-
+        {{default "{{.ROOT_DIR}}/lint-clang-tidy" .OUTPUT_DIR}}
+    requires:
+      vars: ["VENV_DIR"]
+    dir: "{{.ROOT_DIR}}"
+    cmds:
+      - task: ":clang-tidy-output"
+        vars:
+          FLAGS:
+            ref: ".FLAGS"
+          OUTPUT_DIR: "{{.OUTPUT_DIR}}"
+          SOURCE_PATHS:
+            ref: ".DIFF_SOURCE_PATHS"
+          VENV_DIR: "{{.VENV_DIR}}"
+
+  # Runs `clang-tidy` on files found after searching and filtering. Each path in `ROOT_PATHS` is
+  # recursively searched for files. Paths matching any `EXCLUDE_PATTERNS` are skipped. File paths
+  # matching at least one entry from `INCLUDE_PATTERNS` and at least one entry from
+  # `EXCLUDE_PATTERNS` are passed to `clang-tidy`.
+  #
+  # General parameters
+  # @param {string} VENV_DIR Python virtual environment where `clang-tidy` is installed.
+  # @param {string[]} ROOT_PATHS Paths recursively searched for files to run on.
+  # @param {string[]} [FLAGS] Any flags to pass to `clang-tidy`.
+  # @param {string} [OUTPUT_DIR="{{.ROOT_DIR}}/lint-clang-tidy"] Directory to store the `clang-tidy`
+  # output for each file.
+  #
+  # Pattern parameters
+  # @param {string[]} [EXCLUDE_PATTERNS] Wildcard patterns where any matching paths (relative to
+  # `ROOT_PATHS`) will be excluded. The search will not descend into matching directory paths. By
+  # default no paths match.
+  # @param {string[]} [EXTENSION_PATTERNS=["*.cpp", "*.h", "*.hpp"]] Wildcard patterns matched
+  # against the file name (not the full path) determining files to run on.
+  # @param {string[]} [INCLUDE_PATTERNS] Wildcard patterns where any matching file paths (relative
+  # to `ROOT_PATHS`) will be included. By default all paths match.
+  clang-tidy-find:
```

Comment:
Skipped reviewing since it seems similar to the clang-format task and will likely require similar refactoring.